### PR TITLE
rewrite: switch to nushell

### PR DESCRIPTION
**Pull request by oklopfer**
should have a pacstall-programs merge that adds `pacdeps=("nu-tongues-bin")` at same time this is merged as well

### CODE REVIEW

File: rhino-pkg [hunk 1]
```diff
@@ -1,428 +1,531 @@
...
+}
+
+def cmd-exist [input: string] -> bool {
+    let stuff = (which $input)
+    if ($stuff | is-empty) { return false } else if ($stuff).type.0 == "external" { return true }
+}
+
+def search-apt [input: string, desc: bool] -> table {
+    if (cmd-exist 'apt-cache') {
+      let first_table =  if $desc == true {
```
Comment (Elsie19) [inline rhino-pkg:13]:
```suggestion
      let first_table = if $desc == true {
```

File: rhino-pkg [hunk 2]
```diff
@@ -1,428 +1,531 @@
...
+#            #print $inst_pkg
+#            if $package == $inst_pkg { 
+#            let policy_table = ^apt-cache policy $package | detect columns --skip 1 --no-headers
+#            $repo_table.$i.installed = $"(ansi green)\(($policy_table.column1.0)\)(ansi reset)"
+#            break
+#        } }
+       
+            
+        
+
```
Comment (Elsie19) [inline rhino-pkg:42]: Delete these lines.

File: rhino-pkg [hunk 3]
```diff
@@ -1,428 +1,531 @@
...
+#    }
+
+#    repo_table
+
+#}
+
+def prune-search-table [prune_term: string] -> table {
+    let input_table: table = $in
+    let downcase_prune_term = ($prune_term |str downcase)
+    $input_table | filter { |row|    (($row.package | into string | str downcase | str contains $downcase_prune_term) or ($row.description | into string | str downcase | str contains $downcase_prune_term))}
```
Comment (Elsie19) [inline rhino-pkg:52]:
```suggestion
    $input_table | filter { |row| (($row.package | into string | str downcase | str contains $downcase_prune_term) or ($row.description | into string | str downcase | str contains $downcase_prune_term)) }
```

File: rhino-pkg [hunk 4]
```diff
@@ -1,428 +1,531 @@
...
+}
+
+def search-pacstall [input: string] -> table {
+    if (cmd-exist 'pacstall') {
+        let pacstall = do { ^pacstall -S $input  } | complete
```
Comment (Elsie19) [inline rhino-pkg:57]:
```suggestion
        let pacstall = do { ^pacstall -S $input } | complete 
```

File: rhino-pkg [hunk 5]
```diff
@@ -1,428 +1,531 @@
...
+    let output :string = ((($pipeline + ($input | into string)) | str replace "\n" " »|« "))
+    let deansi_output = ($output | ansi strip)
+    let num_of_ansis: int = ($output | split row "\e" | length) - 1
+    let ansi_char_difference: int = ($output | str length --grapheme-clusters) - ($deansi_output | str length --grapheme-clusters)
+
+    mut terminal_width = 100;
+    if (cmd-exist 'tput') {
+        $terminal_width = ((tput cols) | into int)
+    } else if (cmd-exist 'stty') {
+        $terminal_width = ((stty size| split column " ").column2.0 | into int)
```
Comment (Elsie19) [inline rhino-pkg:113]:
```suggestion
        $terminal_width = ((stty size | split column " ").column2.0 | into int)
```

File: rhino-pkg [hunk 6]
```diff
@@ -1,428 +1,531 @@
...
+    # print "Searching flatpak…"
+    translation-dir-path | translate searching.flatpak | print
+    let flatpak = (search-flatpak $input $desc)
+    print -n $"\e[A\e[K"
+    # print "Searching snap…"
+    translation-dir-path | translate searching.snap | print
+    let snap = (search-snap $input)
+    print -n $"\e[A\e[K"
+    #print $extra_prune_terms
+    mut results =  ($apt | append $pacstall | append $flatpak | append $snap )
```
Comment (Elsie19) [inline rhino-pkg:142]:
```suggestion
    mut results = ($apt | append $pacstall | append $flatpak | append $snap )
```

File: rhino-pkg [hunk 7]
```diff
@@ -1,428 +1,531 @@
...
+        # making search terms into one string
+        $search_term += " " 
+        $search_term += $prune_term
+    }
+
+
+    # print -n $"\e[A\e[K"
+    if ($results | is-empty) {
+        # print -e $"No packages found matching '($input)'!"
+        translation-dir-path | translate none-matching {search: $search_term, color: "magenta" } | print
```
Comment (Elsie19) [inline rhino-pkg:159]:
```suggestion
        translation-dir-path | translate none-matching { search: $search_term, color: "magenta" } | print
```

File: rhino-pkg [hunk 8]
```diff
@@ -1,428 +1,531 @@
...
+        exit 1
+    }
+    let results_len = $results | length
+
+    (translation-dir-path | translate found-matching {matches: $results_len, search: $search_term, color: "magenta"} ) + "\n" | print
```
Comment (Elsie19) [inline rhino-pkg:164]:
```suggestion
    (translation-dir-path | translate found-matching { matches: $results_len, search: $search_term, color: "magenta"} ) + "\n" | print
```

File: rhino-pkg [hunk 9]
```diff
@@ -1,428 +1,531 @@
...
+    }
+    mut count = 0
+    # Loop over results
+    for $i in $results {
+        let style = match $i.provider {
+            "pacstall" => $"(ansi yellow_bold)",
+            "apt" => $"(ansi green_bold)",
+            "flatpak" => $"(ansi cyan_bold)",
+            "snap" => $"(ansi red_bold)",
+
```
Comment (Elsie19) [inline rhino-pkg:186]:
```suggestion
```

File: rhino-pkg [hunk 10]
```diff
@@ -1,428 +1,531 @@
...
-    else
-        local contents=("$(apt-cache search "$*" | awk '{print $1}')")
-    fi
-    if [[ -n $contents ]]; then
-        echo "${contents[@]}"
-    else
-        return 1
-    fi
+
+
```
Comment (Elsie19) [inline rhino-pkg:265]:
```suggestion
```

File: rhino-pkg [hunk 11]
```diff
@@ -1,428 +1,531 @@
...
+def user-package-selection [ input: table , purpose: string ] -> table {
+    mut user_input = ""
+    print ""
+    let input_final_index = ($input | length) - 1
+
+   $user_input = match $purpose {
+        "install" => (input (translation-dir-path| translate ask.which-install {index: $input_final_index}) | into string),
```
Comment (Elsie19) [inline rhino-pkg:272]:
```suggestion
        "install" => (input (translation-dir-path | translate ask.which-install { index: $input_final_index }) | into string),
```

File: rhino-pkg [hunk 12]
```diff
@@ -1,428 +1,531 @@
...
+        
+        "info" => (input (translation-dir-path |  translate ask.which-info{ index: $input_final_index }) |into string),
+        "remove" => (input (translation-dir-path| translate ask.which-remove {index: $input_final_index}) | into string),
```
Comment (Elsie19) [inline rhino-pkg:275]:
```suggestion
        "remove" => (input (translation-dir-path | translate ask.which-remove { index: $input_final_index }) | into string),
```

File: rhino-pkg [hunk 13]
```diff
@@ -1,428 +1,531 @@
...
+
+def user-package-selection [ input: table , purpose: string ] -> table {
+    mut user_input = ""
+    print ""
+    let input_final_index = ($input | length) - 1
+
+   $user_input = match $purpose {
+        "install" => (input (translation-dir-path| translate ask.which-install {index: $input_final_index}) | into string),
+        
+        "info" => (input (translation-dir-path |  translate ask.which-info{ index: $input_final_index }) |into string),
```
Comment (Elsie19) [inline rhino-pkg:274]:
```suggestion
        "info" => (input (translation-dir-path |  translate ask.which-info { index: $input_final_index }) |into string),
```

File: rhino-pkg [hunk 14]
```diff
@@ -1,428 +1,531 @@
...
+        _ => []
+    }
+
+    #uses regex to filter out non-number inputs, then converts to int
+    mut user_input_ints = ($user_input | split row ' ' | find --regex "[0-9]+" | find --regex "^[0-9]+" | into int) 
+    
+    #screens the list for invalid indices
+    mut drop_list: list<int> = []
+    for i in 0..<($user_input_ints | length) {
+        if ($user_input_ints |select $i).0 > $input_final_index {
```
Comment (Elsie19) [inline rhino-pkg:285]:
```suggestion
        if ($user_input_ints | select $i).0 > $input_final_index {
```

File: rhino-pkg [hunk 15]
```diff
@@ -1,428 +1,531 @@
...
+            $drop_list = ($drop_list | append $i)
+        }  
+    }
+    #prunes invalid indices
+    for i in 0..<($drop_list|length) {
```
Comment (Elsie19) [inline rhino-pkg:290]:
```suggestion
    for i in 0..<($drop_list | length) {
```

File: rhino-pkg [hunk 16]
```diff
@@ -1,428 +1,531 @@
...
+        $user_input_ints = ($user_input_ints | drop nth ($drop_list| select $i).0)
```
Comment (Elsie19) [inline rhino-pkg:291]:
```suggestion
        $user_input_ints = ($user_input_ints | drop nth ($drop_list | select $i).0)
```

File: rhino-pkg [hunk 17]
```diff
@@ -1,428 +1,531 @@
...
+    }
+
+    #user provided no numbers that were valid
+    if ($user_input_ints | is-empty) {
+        let error_msg = translation-dir-path | translate invalid.integers {number: $input_final_index}
```
Comment (Elsie19) [inline rhino-pkg:296]:
```suggestion
        let error_msg = translation-dir-path | translate invalid.integers { number: $input_final_index }
```

File: rhino-pkg [hunk 18]
```diff
@@ -1,428 +1,531 @@
...
+        error make {msg: $error_msg}
```
Comment (Elsie19) [inline rhino-pkg:297]:
```suggestion
        error make { msg: $error_msg }
```

File: rhino-pkg [hunk 19]
```diff
@@ -1,428 +1,531 @@
...
+
+    $user_input_ints | each { |index|
+        let pkg = ($input | get $index | get package)
+        let provider = ($input | get $index | get provider)
+        
+        
+        #print $"Selecting '(ansi purple_bold)($pkg)(ansi reset)' from package manager '(ansi purple_bold)($provider)(ansi reset)'"
+        
+        if $install {
+            translation-dir-path | translate install-select {package: $pkg, manager: $provider, color: "magenta" } | print
```
Comment (Elsie19) [inline rhino-pkg:316]:
```suggestion
            translation-dir-path | translate install-select { package: $pkg, manager: $provider, color: "magenta" } | print
```

File: rhino-pkg [hunk 20]
```diff
@@ -1,428 +1,531 @@
...
+        } else {
+            translation-dir-path | translate remove-select {package: $pkg, manager: $provider, color: "magenta" } | print
```
Comment (Elsie19) [inline rhino-pkg:318]:
```suggestion
            translation-dir-path | translate remove-select { package: $pkg, manager: $provider, color: "magenta" } | print
```

File: rhino-pkg [hunk 21]
```diff
@@ -1,428 +1,531 @@
...
+        }
+
+
+        let r_u_sure = translation-dir-path| translate ask.sure 
+        let sure = (input --numchar 1 $"($r_u_sure) ")
+        
+        let no: bool = (($sure != "Y") and ($sure != "y"))
+        # let sure = (input $"Are you sure? \((ansi green_bold)y(ansi reset)/(ansi red_bold)N(ansi reset)\) ")
+        if $no {
+            translation-dir-path | translate skipping {package: $pkg, manager: $provider, color: "magenta"} | print
```
Comment (Elsie19) [inline rhino-pkg:328]:
```suggestion
            translation-dir-path | translate skipping { package: $pkg, manager: $provider, color: "magenta" } | print
```

File: rhino-pkg [hunk 22]
```diff
@@ -1,428 +1,531 @@
...
-        local contents=("$(snap find "$*" | awk '{ print $1 }' | tail -n +2)")
-    fi
-    if [[ ${contents[*]} == "No matching snaps for"* ]]; then
-        return 1
-    else
-        echo "${contents[@]}"
-    fi
+def info [ ] {
+    let input: table = $in
+    let user_input_ints =  user-package-selection $input "info"
```
Comment (Elsie19) [inline rhino-pkg:354]:
```suggestion
    let user_input_ints = user-package-selection $input "info"
```

File: rhino-pkg [hunk 23]
```diff
@@ -1,428 +1,531 @@
...
-    pacstall_cmd="-R"
-    prompt $"Select which package to remove" "$count"
-fi
-
-read -ra entered_input
-echo -ne "${NC}"
-if [[ ! ${entered_input[*]} =~ ^(([0-9])\s?)+ ]]; then
-    msg $"'${entered_input[*]}' is not a valid number"
+
+def search-cmd-no-args [ ] {
```
Comment (Elsie19) [inline rhino-pkg:360]:
```suggestion
def search-cmd-no-args [] { 
```

File: rhino-pkg [hunk 24]
```diff
@@ -1,428 +1,531 @@
...
-fi
-
-read -ra entered_input
-echo -ne "${NC}"
-if [[ ! ${entered_input[*]} =~ ^(([0-9])\s?)+ ]]; then
-    msg $"'${entered_input[*]}' is not a valid number"
+
+def search-cmd-no-args [ ] { 
+    let command: string = $in
+    let error_msg = translation-dir-path | translate invalid.search-arguments {subcommand: $command}
```
Comment (Elsie19) [inline rhino-pkg:362]:
```suggestion
    let error_msg = translation-dir-path | translate invalid.search-arguments { subcommand: $command }
```

File: rhino-pkg [hunk 25]
```diff
@@ -1,428 +1,531 @@
...
+    error make -u {msg: $error_msg}
```
Comment (Elsie19) [inline rhino-pkg:363]:
```suggestion
    error make -u { msg: $error_msg }
```

File: rhino-pkg [hunk 26]
```diff
@@ -1,428 +1,531 @@
...
+        "search" | "srch" | "find" | "s" => "search",
+        "update" | "updt" | "upgrade" | "upd" | "upg" | "u"  => "update",
+        "cleanup" | "clnp" | "cln" | "clean" | "c" => "cleanup",
+        
+        _ => "invalid"
+        
+   }
+# "info" | "information" | "data" | "query" | "qry" | "q" => "info",
+    if $command == "invalid" {
+        let error_msg = translation-dir-path | translate invalid.subcommand {subcommand: $rest.0}
```
Comment (Elsie19) [inline rhino-pkg:469]:
```suggestion
        let error_msg = translation-dir-path | translate invalid.subcommand { subcommand: $rest.0 }
```

File: rhino-pkg [hunk 27]
```diff
@@ -1,428 +1,531 @@
...
+    }
+
+    match $command {
+        "update" => (update $yes),
+        "cleanup" => (cleanup $yes)
+    }
+
+
+
+
```
Comment (Elsie19) [inline rhino-pkg:529]: Remove these.

**Review by Elsie19 — CHANGES_REQUESTED** [review]

**Comment by oklopfer** [comment]
Makefile will also need to be edited for the tomls

File: rhino-pkg [hunk 28]
```diff
@@ -1,428 +1,531 @@
...
-             the number(s) associated with the desired package(s).
-             
-    search:  Search for package(s) - Does not have a second prompt.
-    
-    update:  Updates all packages accessible to the wrapper - does
-             not accept <input>, instead use install to update 
-             individual packages. Has a confirmation prompt.
+#!/usr/bin/env nu
+def translation-dir-path [] -> string {
+    "/usr/src/pacstall/rhino-pkg/translation_tomls"
```
Comment (oklopfer) [inline rhino-pkg:3]: this is not a valid dir that will ever exist. it should be in a newly created /usr/local/share/rhino-pkg, not /usr/src/pacstall

File: rhino-pkg [hunk 29]
```diff
@@ -1,428 +1,531 @@
...
+#     'coooooooo:oo.';.
+#   ,oooooooooooooooo    ;
+#  clllcccllloooooooo;c:'o
+# .;';:::::::::cclooooooo'
+# ''',::::::::::::::ccclc.
+# .''';::::::::::l:::::::
+#  '''',:::::::::kd.
+#  .''''',;::ck:oW;
+#    ''''''''kXOM.
+#      .,,:dXMK
```
Comment (oklopfer) [inline rhino-pkg:437]: this page lost all of its color when displaying in terminal (which is actually quite important)... can we bring it back?

Comment (Elsie19) [reply rhino-pkg:437]: No.

Comment (oklopfer) [reply rhino-pkg:437]: Then it should be a manpage instead, without the color it’s confusing and hard to read

Comment (Elsie19) [reply rhino-pkg:437]: Manpages can't be colored with ansi codes.

**Comment by Elsie19** [comment]
@oklopfer any updates on the changes I suggested?

**Comment by oklopfer** [comment]
Yeah, you haven’t done any of them yet. It’s your project, Elsie. You maintain it. Not me.

**Comment by Elsie19** [comment]
It's your PR but fine.

**Comment by oklopfer** [comment]
It’s also all @wren54 ‘s commits. I just made the PR so we could track it.

**Comment by Elsie19** [comment]
This one isn't happening, but look at https://github.com/Elsie19/rpk2

**Pull request closed without merging**